Implement `jest-haste-map` instead of `node-haste`

cpojer opened the pull request:
This is a new haste map implementation for Jest which is much more scalable than node-haste.

node-haste2 isn't well designed and not scalable for short-lived services like Jest. The startup time of node-haste1 vs. node-haste2 on www is almost the same, both between 6 and 8 seconds which is not acceptable for our engineers. This implementation is attempting to accomplish a much reduced and more scalable startup time. It also has reduced scope – the goal of this is to only build a haste map and provide a way to resolve a one-level deep dependency tree, which is all that Jest really needs from node-haste. `jest-haste-map` can serve as an ideal basis for rewriting node-haste2 (into node-haste3!).

With this, the cold start time (building the entire haste map) is now about 14 seconds on www (that's acceptable) but the incremental invocation is only 2 seconds (@kyldvs will love me). I haven't heavily micro-optimized the JavaScript in `packages/jest-haste-map/src/index.js` and there is a bunch of data copying with `HasteResolver.js` – I believe I can get close to 1 second with these optimizations once I'm done. One of the goals I have is to allow tacking on data (list of mocks) to the haste map and serialize the haste map and read that one in directly in the workers instead of keeping two caches.

On every startup node-haste2 asks watchman (or node) for a list of *all* files. It builds up an in memory-file system and then processes every single file through an insane amount of promise chaining. It takes about 3 seconds on react-native to do this on a warm start. With this new implementation the cold start is about 1.5 seconds and warm start is way less than a second.

`jest-haste-map` only does the minimum processing necessary: it retrieves a list of all files only on the first start when using watchman and subsequent invocations only receive the deltas of changed files since the last time Jest has run. It further processes only files that have been changed or added. The node crawler isn't as optimized but it should be faster than node-haste2 as well.

What to review:
* `packages/jest-haste-map/src/index.js`
* `packages/jest-haste-map/src/worker.js`
* `packages/jest-haste-map/src/crawlers/`
* The other modules in `jest-haste-map` were lifted from node-haste2 and just received small code style updates (removed babel compilation etc.)
* `src/resolvers/HasteResolver.js` for usage.

cc @davidaurelio @kentaromiura @jeffmo @vjeux
cc @amasad – if you could take a quick look, I would greatly appreciate it :)

cpojer commented:
Actually the time it takes to run a fast test (~10ms) on www with this is 2s overall. That includes the 500ms overhead of loading jsdom too, so we are already pretty close to 1s.

cpojer commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,57 @@
+    path.sep + '**' + path.sep + '*.{' + extensions.join(',') + '}';
+
+  return Promise.all(roots.map(
+    root => globPromise(root + pattern)
```
this nice and simple implementation of the node crawler actually turns out to have 2x worse performance than the one in node-haste. I might revert to that one instead. Didn't think glob would be sooo slow.

cpojer commented:
Done in e3189c4c80109e60f42cd98fa94880066587df9b: "Clean up HasteResolver and file caching for all of this". Now we only create one cache file. It is amended by "HasteResolver" with mocking information and is directly read from a TestWorker.

One nice side-effect of all this work is that Jest's own test suite is almost 2x as fast now. It only takes 2.5s instead of 4-5s.

davidaurelio commented on packages/jest-haste-map/src/__tests__/fastpath-test.js:
```diff
@@ -0,0 +1,527 @@
+/*
+ * original author: dead_horse <dead_horse@qq.com>
+ * ported by: yaycmyk <evan@yaycmyk.com>
+ */
```
It would be nice if we could take over the fastpath name on npm :-)

janicduplessis replied:
"fasterpath" :)

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,83 @@
+function nodeCrawl(roots, extensions, ignorePattern, data) {
+  const extensionsPattern = new RegExp('\.(' + extensions.join('|') + ')$');
+  const files = Object.create(null);
+  const queue = roots.slice();
```
do you remember the case for multiple roots?

cpojer replied:
In Jest at least, we have `testPathDirs` and each of them is a root. On www it is more practical to crawl certain folders rather than all of them.

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,83 @@
+          const name = fileData[0];
+          const stat = fileData[1];
+          // Ignore broken links.
+          if (!stat || ignorePattern.test(name)) {
```
is it worth filtering with `ignorePattern` before `stat`ing?

cpojer replied:
yesss, also the check for extensions, I missed that! :)

davidaurelio commented on packages/jest-haste-map/src/crawlers/watchman.js:
```diff
@@ -0,0 +1,107 @@
+const watchmanURL = 'https://facebook.github.io/watchman/docs/troubleshooting.html';
+
+function isDescendant(root, child) {
+  return child.startsWith(root);
```
does `root` always end with `path.sep`?

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,241 @@
+    const read = this._options.resetCache ? this._createEmptyMap : this.read;
+
+    return Promise.resolve()
+      .then(() => read.call(this))
```
`Promise.resolveI(read.call(this))`

or

`Promise.resolve(this._options.resetCache ? this._createEmptyMap() : this.read())`

cpojer replied:
that doesn't work the way I want I don't think. I want the error to be thrown inside the promise so that it can be caught with `catch`.

stryju replied:
and the overhead created via `bind` is negligible, no?

davidaurelio replied:
> and the overhead created via bind is negligible, no?

not in v8. but don’t ask me for numbers

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,241 @@
+    return readFile(filePath, 'utf-8')
+      .then(data => {
+        data = JSON.parse(data);
+        if (data.name) {
```
Isn’t it a package if there is no name given? Node [doesn’t care about the name](https://nodejs.org/api/modules.html#modules_folders_as_modules): should we default to the directory?

cpojer replied:
This is only really for haste-packages which is for package.json files that are not part of a node_modules folder. Jest does its own resolution, so we don't need to read all packages. The haste packages all have a name, so this should be fine.

davidaurelio commented:
congrats on the speed improvements, that’s awesome!

I don’t have too much to say about the code. The indentation level of promises can also get quite deep :-)
What was our case for multiple roots? Do we really need that?

davidaurelio commented:
Should we write a dex with the haste resolution rules?

davidaurelio commented:
we should try to get the “fastpath” module and npm and publish the two file crawlers seperately + an watchman auto-detection function.

davidaurelio commented:
A quick synthetic benchmark seems to suggest that native `Map` and `Set` are really fast in node.js, even faster than property lookups on objects with many entries. So there isn’t really a case for `Object.create(null)`. Deserialization from JSON also doesn’t need the `setPrototypeOf` trick. Serialization, on the other hand, is uglier (`.stringify(Array.from(map.entries()))`)

cpojer commented:
Thanks for the review.

* We use `testPathDirs` as separate roots extensively on www. This reduces the scope of files to crawl
* Yes, we should probably write a dex :D 
* It is probably fine to publish this as `jest-haste-map` as a first version and then consider splitting out certain packages into separate modules, depending on what other people might need.
* Yes, Map and Set are super fast but they don't serialize well to JSON. Is writing to a Map or Set as fast as it is to an Object? We are doing so many writes during startup, which is quite critical when you have 100k+ items. https://github.com/cognitect/transit-js could help to speed up serialization – it is basically a custom serializer that is really fast and can encode/decode to immutable data structures. I think right now that is more than I need but if it yields real performance gains, I'd consider going for it :)

cpojer commented:
I made a bunch of micro-optimizations in ed2ee7829928f890cb3c1792d1b9d0cf93f9061a

Since both my metadata objects for all files/modules contain the same fixed amount of items, using constant keys and arrays  was a great optimization and the code complexity stayed pretty much the same (`module[H.ID]` instead of `module.id`, for example).

The file size reduction of using constant keys and arrays instead of long string names and objects was 10 mb (!) down from 26 to 16 on a large code base. I was then able to exclude 5 % of the files from the original list and am now down to 14 mb. The parse time went down from 500ms to 250ms. The write time, interestingly, is still at 300ms. I looked into compacting the paths but that would only yield a win of another 2mb or so but it would add additional post-processing. I believe with this model the file size will grow to 30 mb over the next 3-4 years and decode in less than a second which will be more than good enough. Then, someone can consider different serialization models but right now we are in the diminishing-returns territory. The data model is simple enough and the code is easy enough to change.

Things I tried:
* JSONStream: this is memory efficient but actually very slow
* Decoding every metadata object individually so I could store them using my own data format. `loop over JSON.parse(item)` had *exactly* the same performance characteristics as `JSON.parse(allItems);`.

stryju commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,268 @@
+class HasteMap {
+
+  constructor(options) {
+    this._options = {
```
why not `Object.assign`?

```
this._options = Object.assign({
  cacheDirectory: os.tmpdir(),
  mocksPattern: options.mocksPattern
    ? new RegExp(options.mocksPattern)
    : null,
  useWatchMan: true,
};
```

---

also - shouldn't this be `os.tmpdir()` instead of `os.tmpDir()`?

rivertam replied:
I think you forgot a crucial part of this example:

this._options = Object.assign({
  cacheDirectory: os.tmpdir(),
  mocksPattern: options.mocksPattern
    ? new RegExp(options.mocksPattern)
    : null,
  useWatchMan: true,
}, options);

cpojer replied:
I guess I could do that.

Also: https://github.com/nodejs/node/blob/master/lib/os.js#L48 lol nodejs.

cpojer replied:
Actually I do like the full enumeration of all configuration options.

cpojer commented:
This implementation is now done and ready for review. I've added tests for all but the main `HasteMap` file which I will do tomorrow. Jest works on all platforms I have tested (www, react-native) so I'm confident the implementation is correct.

I added a large docblock to `HasteMap` explaining the constraints and architecture. I will add more documentation to individual methods.

On www at Facebook these are the effects of the change (very large repo):
* way more predictable performance than node-haste – variance is less than a second on repeat runs.
* cold start (no cache): 20-30s down to 6s
* warm start: 6s down to <1s (this was the goal)
* inverse dependency resolution: 12-17s down to 0.3s

cpojer commented on .eslintrc:
```diff
@@ -9,7 +9,6 @@
     "arrow-spacing": [2],
     "babel/arrow-parens": [2, "as-needed"],
     "brace-style": [2, "1tbs", {"allowSingleLine": true}],
-    "camelcase": [2, {"properties": "always"}],
```
I added this recently but it is really silly, especially when working with data from an API I don't control. Let's remove it and be less strict, yay!

davidaurelio replied:
I know that situation. I used to disable it where necessary. There’s also the [possibility to disable it for property names](http://eslint.org/docs/rules/camelcase#never).

lencioni commented on packages/jest-haste-map/src/constants.js:
```diff
@@ -0,0 +1,34 @@
+  DEPENDENCIES: 3,
+
+  /* module map attributes */
+  PATH: 0,
```
The fact that `PATH === ID` etc makes me a little uneasy.

cpojer replied:
well they are in different arrays so it shouldn't be a big deal. One nice way to think about this is that the first key in either the files or map object can be used to access the corresponding metadata object in the other :)

lencioni commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+            if (stat.isDirectory()) {
+              search(file);
+            } else {
+              const ext = path.extname(file).substr(1);
```
nit: `slice` is a bit nicer than `substr` IMHO

lencioni commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+      args.push('-o');
+    }
+    args.push('-iname');
+    args.push('*.' + ext);
```
nit: template?

davidaurelio replied:
```
args.push('*.' + ext);
args.push(`*.${ext}`);
```

same length. I thing the first one is clearer.

lencioni commented on packages/jest-haste-map/src/lib/extractRequires.js:
```diff
@@ -0,0 +1,41 @@
+/* eslint-disable max-len */
+const replacePatterns = {
+  EXPORT_RE: /(\bexport\s+(?:[^'"]+\s+from\s+)??)(['"])([^'"]+)(\2)/g,
+  IMPORT_RE: /(\bimport\s+(?:[^'"]+\s+from\s+)??)(['"])([^'"]+)(\2)/g,
```
These aren't quite right. `[^'"]` will stop at either `'` or `"`, which is not what you want. In practice, this seems unlikely to be a problem for these statements, but I think you should use `.+?` instead.

cpojer replied:
I honestly just copied this from node-haste and am scared of touching these regexes! Where do you believe would this be an issue? :)

davidaurelio replied:
these also follow `import type x from`

while we’re at it, consider
```
// not a type import
import type from "x"; 

// type imports
import type x from "x";
import type {x} from "x";
import type from from "x";
import type type from "x";
```

lencioni replied:
For instance, if you have a string like this:

```js
import foo from 'foo"bar';
```

the regex here won't match. If you use this instead:

```js
/(\bimport\s+(?:[^'"]+\s+from\s+)??)(['"])(.+?)(\2)/g
```

it will match. As I said, in practice it is probably fine, but this seems better.

```diff
- /(\bimport\s+(?:[^'"]+\s+from\s+)??)(['"])([^'"]+?)(\2)/g
+ /(\bimport\s+(?:[^'"]+\s+from\s+)??)(['"])(.+?)(\2)/g

davidaurelio replied:
> the regex here won't match. If you use this instead:
> ```js
/(\bimport\s+(?:[^'"]+\s+from\s+)??)(['"])(.+?)(\2)/g
```

Well, that won’t match 
```js
import foo from 'foo\'bar';
```

dmitriiabramov commented on src/TestRunner.js:
```diff
@@ -47,11 +57,19 @@ class TestRunner {
+    this._hasteMap.build().then(data => {
+      console.timeEnd('build');
+      console.log('files', Object.keys(data.files).length);
+      console.log('modules', Object.keys(data.map).length);
```
i assume that was for debugging? They print a lot :)

cpojer replied:
yeah, I'll remove this :)

cpojer commented:
And we are done. I tried to document the haste map as much as I could and everything should be covered by tests. Jest itself is an integration test for `jest-haste-map`. Everything works well on all code bases I tried. Jest is much faster for really really tiny repos (no more startup cost) and much faster for huge repos. Same perf for mid-sized libraries (like relay).

cpojer commented:
@facebook-github-bot import

davidaurelio commented on packages/jest-haste-map/src/__tests__/fastpath-test.js:
```diff
@@ -62,6 +64,7 @@ describe('fast-path', () => {
     it('should resolve given paths', () => {
       process.platform = actualPlatform;
 
+      fp = require('../fastpath');
```
`beforeEach`

davidaurelio commented on packages/jest-haste-map/src/__tests__/fastpath-test.js:
```diff
@@ -0,0 +1,575 @@
+    it('should resolve given paths', () => {
+      process.platform = actualPlatform;
+
+      fp = require('../fastpath');
```
move all of these to `beforeEach` maybe?

cpojer replied:
I can't because we modify `process.platform` in each test and the module depends on it in the module factory. We wipe it away on every test, change the platform and freshly require fastpath.

The reason why this works differently in the node-haste repo right now is because we use the inline-requires transform there. This repo doesn't have a build step, so I'm just doing the requires manually :)

davidaurelio commented on packages/jest-haste-map/src/__tests__/fastpath-test.js:
```diff
@@ -0,0 +1,575 @@
+      expect(fp.extname('file.\\\\')).toEqual('.');
+    });
+
+    it('should handle path backslashes (posix)', () => {
```
don’t we need to set `process.platform = 'linux'` here to make these tests pass on windows?

cpojer replied:
we actually always default to linux in `beforeEach`, so this should be safe :)

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+
+function find(roots, extensions, ignore, callback) {
+  const result = [];
+  let activeCalls = 0;
```
async/queue would make this clearer and less error prone imho.

cpojer replied:
I agree with you. I lifted this from node-haste1 and only updated the code style to fit ES2015 better. I figured I'll leave it as is so I don't break it :)

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+function findNative(roots, extensions, ignore, callback) {
+  const args = [].concat(roots);
+  args.push('-type', 'f');
+  extensions.forEach((ext, index) => {
```
you need parentheses around the extensions

`find x -type f -name '*.json' -o -name '*.js'` is not equivalent to
`find x -type f \( -name '*.json' -o -name '*.js' \)`

cpojer replied:
wait, what's the difference? It seems to work fine on www for example. This is what it was like in node-haste1.

davidaurelio replied:
the first one would also match directories with ending with “.js”

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+  roots.forEach(search);
+}
+
+function findNative(roots, extensions, ignore, callback) {
```
nice optimization :+1:

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+  const child = spawn('find', args);
+  let stdout = '';
+  child.stdout.setEncoding('utf-8');
+  child.stdout.on('data', data => stdout += data);
```
concat-stream allows you not to write this code ;-)

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+          result.push([path, stat.mtime.getTime()]);
+        }
+        if (--count === 0) {
+          callback(result);
```
async/parallel

davidaurelio commented on packages/jest-haste-map/src/crawlers/node.js:
```diff
@@ -0,0 +1,121 @@
+        if (existingFile && existingFile[H.MTIME] === mtime) {
+          files[name] = existingFile;
+        } else {
+          files[name] = ['', mtime, 0, []];
```
I’d add a comment pointing to constants.js here

cpojer replied:
good point!

davidaurelio commented on packages/jest-haste-map/src/crawlers/watchman.js:
```diff
@@ -0,0 +1,112 @@
+          // files.
+          if (!reset && response.is_fresh_instance) {
+            reset = true;
+            files = Object.create(null);
```
If I understand the code correctly, this could break when using multiple roots:

if the first response has `is_fresh_instance == false`, but it’s `true` for the second, we would lose the data of the first response. Is that correct?

cpojer replied:
yes, that is actually true. The only way to avoid this would be to wait for all queries to finish, then check if one of them is fresh and reset the files object. This might be slower as we'd wait for the slowest operation to finish until we can continue. However, multiple roots doesn't equal multiple watchman roots. In all cases for Facebook (and Jest), all the passed in roots reduce to a single watchman root. Therefore waiting for all queries is not a problem, because it is always just one. I'll make this change.

cc @wez: can it ever happen that watchman gives `is_fresh_instance` for one root but not another? Or will watchman recrawl every watched project if it misses notifications from the file system? I'm assuming when watchman restarts, `is_fresh_instance` would be true for all projects anyway?

davidaurelio replied:
you could create a `Map` per root and combine them at the end.

wez replied:
@cpojer Each distinct watch has its own set of buffers and can thus be started, restarted or overflown and yield a `fresh_instance` result set independently.

watches established via `watch-project` can consolidate and re-use an existing watch; for example, if you were to `watch-project www/some/dir` it would re-use the main watch on `www` and then query the sub directory.  If you have multiple clients or subscriptions focused on this kind of watch, they would all receive a `fresh_instance` result set if or when the underlying watch starts/restarts/overflows.

If the reason for generating a `fresh_instance` result set was an overflow due to excessive system load, there is a good chance that it will impact each distinct watch at around the same time (no guarantees on timing because this is an inherently racy space), possibly cascading from one to the other if those watches are all being processed by a single thread and it gets swamped.

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+  static getCacheFilePath(tmpdir) {
+    const hash = crypto.createHash('md5');
+    Array.from(arguments).slice(1).forEach(arg => hash.update(arg));
+    return path.join(tmpdir, hash.digest('hex'));
```
it would be nice to add a human-readable prefix here. That makes it easier to delete these files manually if necessary for debugging.

cpojer replied:
good idea! Let's use the name from the options and make it `tmpdir/<name>-hash`.

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+    if (!this._buildPromise) {
+      this._buildPromise = this._buildFileMap()
+        .then(fileMap => this._buildHasteMap(fileMap))
+        .then(hasteMap => this._persist(hasteMap));
```
Does persisting need to be sync and blocking at all? There isn’t really a need to wait for it, or is there?

cpojer replied:
In Jest all the worker processes need this map for module resolution. I was thinking of making this parallel to some other non I/O bound operations to save time but I didn't want to complicate the `jest-haste-map` API just yet.

davidaurelio replied:
ok, that makes sense. Is reading the file from the worker faster than using ipc to transfer the map?

cpojer replied:
worker-farm doesn't have a way to send initial data to a process (there is no concept of start or stop for an individual worker) so I'd have to send it on every call. We call it 4000 times on www (once for every test) and sending the entire map over definitely takes longer than reading it once per worker. Wish node had shared immutable memory…

davidaurelio replied:
yup. I’m breaking up with worker-farm for packager btw. I need something that can scale up dynamically.

cpojer replied:
It works pretty well for distributed tasks with no state.

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+    return path.join(tmpdir, hash.digest('hex'));
+  }
+
+  build() {
```
Personally, I’d prefer this to be a static method that returns a promise that resolves to an instance. That way, there’s no need to call `this.build().then(…)` / `hasteMap.build().then(…)` anywhere.

But that’s a cosmetic change, so feel free to ignore this comment :-)

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+  /**
+   * 3. parse and extract metadata from changed files.
+   */
+  _buildHasteMap(hasteMap) {
```
nit: `hasteMap` should be `fileMap`

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+      promises.push(
+        this._getWorker()({filePath}).then(
+          metadata => {
+            fileMetadata[H.VISITED] = 1;
```
nit: `true`

cpojer replied:
unfortunately `true` and `false` are much longer when encoding them as strings. I was thinking of adding constants for this instead.

davidaurelio replied:
if it’s an optimization just comment it :)

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+          },
+          error => {
+            // If a file cannot be read we remove it from the file list and
+            // ignore the failure silently.
```
I’m curious: under which circumstances does this become relevant?

cpojer replied:
I believe watchman's `exists` flag on files might not always be entirely correct. I chose to not trust watchman here and just ignore errors silently. We might want to log or throw at some point but it should be good enough for now.

wez replied:
the exists flag is accurate in the sense that it reflects the existence of the file at the time that watchman was notified about it, but you need to keep in mind that filesystem watching is full of TOCTOU issues; since watchman delivered a notification to you about the existence of the file, some other process could have removed it and there may well be a pending notification about that removal that you will receive in a few milliseconds.  Similarly for removal/re-creation.

With that in mind, I'd advocate treating any notification from any file watcher as advisory.

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+
+  _parse(hasteMap) {
+    hasteMap = JSON.parse(hasteMap);
+    for (const key in hasteMap) {
```
Are these arrays now? In that case the `.setPrototypeOf` would be unnecessary

cpojer replied:
nope, they are all Objects. See the docblock on top of this class. The values in these objects are arrays however (for `files` and `map`).

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+    return hasteMap;
+  }
+
+  _crawl(hasteMap) {
```
In order to make this class useful for the packager version that will be integrated with buck, this should be factored out. The same is valid for creating/persisting the cache.

The packager, when driven by buck, will receive a list of files and won’t crawl itself.

cpojer replied:
Are you going to make buck a dependency of packager even for open source? In either case, I think this should be pretty easy to support: the options passed into an instance can support a `crawl` function and for buck supported packager it can be as simple as `crawl: () => Promise.resolve(filesFromBuck);`.

davidaurelio commented on packages/jest-haste-map/src/index.js:
```diff
@@ -0,0 +1,370 @@
+    );
+  }
+
+  _ignore(filePath) {
```
Buck/packager integration: same concern here.

cpojer replied:
we can just make this optional as well. Ignore is only called when crawling though.

davidaurelio commented on packages/jest-haste-map/src/lib/docblock.js:
```diff
@@ -0,0 +1,50 @@
+
+const commentEndRe = /\*\/$/;
+const commentStartRe = /^\/\*\*/;
+const docblockRe = /^\s*(\/\*\*(.|\r?\n)*?\*\/)/;
```
fbsource contains modules that look like this:
```js
/*
 * @providesModule x
 */
```
(no double star)

and also contain `'use strict'` and single line comments before the docblock (although you have that covered)

davidaurelio replied:
```js
const docblockRe = /^\s*(\/\*\*[^]*?\*\/)/;
```
also covers unicode line breaks

cpojer replied:
This is just the same regex we are already using in node-haste. I'm happy to make changes to these in a follow-up but I'm worried of changing behavior in this diff.

davidaurelio commented on packages/jest-haste-map/src/lib/docblock.js:
```diff
@@ -0,0 +1,50 @@
+
+function extract(contents) {
+  const match = contents.match(docblockRe);
+  return match ? match[0].replace(ltrimRe, '') || '' : '';
```
use 
```js
const docblockRe = /^\s*(\/\*\*([^]*?)\*\/)/;
```
and `match[1]`. That will give you the docblock without leading `/**` and trailing `*/`. Also, `|| ''` should be unecessary.

cpojer replied:
same as above, I'm trying not to change anything here.

davidaurelio commented on packages/jest-haste-map/src/lib/docblock.js:
```diff
@@ -0,0 +1,50 @@
+}
+
+function parse(docblock) {
+  docblock = docblock
```
not sure whether you need a full blown parser at all. You’re just interested in `@provides`/`@providesModule`, correct?

Extracting the docblock and then matching against
```js
/^\s*\**\s*@provides(?:Module)[*\s](\S+)/m
```

might be the easier and faster way to get what you want.
Do we really have

```
/**
 * @provides
 * modulename
 */
```
in our code base? That’s scary

cpojer replied:
yes. In www we also have `@requires` with multi-line requires like

```
@requires foo-css
                  foo-js
```

etc.

davidaurelio commented:
This looks solid in general.

To make this the base of node-haste (and have RN packager continue to use node-haste), we need better decoupling of caching, crawling, and building the map.
